mesonbuild/dependencies/pkgconfig.py

```diff
@@ -78,38 +121,30 @@ def list_all(self) -> T.List[str]:
 class PkgConfigCLI(PkgConfigInterface):
     '''pkg-config CLI implementation'''
 
-    # The class's copy of the pkg-config path. Avoids having to search for it
-    # multiple times in the same Meson invocation.
-    class_pkgbin: PerMachine[T.Union[None, Literal[False], ExternalProgram]] = PerMachine(None, None)
-    # We cache all pkg-config subprocess invocations to avoid redundant calls
-    pkgbin_cache: T.Dict[
-        T.Tuple[ExternalProgram, T.Tuple[str, ...], T.FrozenSet[T.Tuple[str, str]]],
-        T.Tuple[int, str, str]
-    ] = {}
-
     def __init__(self, env: Environment, for_machine: MachineChoice, silent: bool) -> None:
-        self.env = env
-        self.for_machine = for_machine
-        # Store a copy of the pkg-config path on the object itself so it is
-        # stored in the pickled coredata and recovered.
-        self.pkgbin = self._detect_pkgbin(env, for_machine, silent)
+        super().__init__(env, for_machine)
+        self._detect_pkgbin()
+        if self.pkgbin and not silent:
+            mlog.log('Found pkg-config:', mlog.green('YES'), mlog.bold(f'({self.pkgbin.get_path()})'), mlog.blue(self.pkgbin_version))
 
     def found(self) -> bool:
         return bool(self.pkgbin)
 
+    @lru_cache(maxsize=None)
     def version(self, name: str) -> T.Optional[str]:
         mlog.debug(f'Determining dependency {name!r} with pkg-config executable {self.pkgbin.get_path()!r}')
         ret, version, _ = self._call_pkgbin(['--modversion', name])
         return version if ret == 0 else None
 
     @staticmethod
-    def _define_variable_args(define_variable: T.Optional[ImmutableListProtocol[str]]) -> T.List[str]:
+    def _define_variable_args(define_variable: PkgConfigDefineType) -> T.List[str]:
```

Comment:
This needs to return an `ImmutableListProtocol[str]` now. Otherwise if a consumer mutates the returned value it mutates the returned value it will mutate the value in the cache

Reply:
done